.github/workflows/build_and_test.yaml

```diff
@@ -16,54 +37,110 @@ on:
 
 jobs:
   build:
-    name: Build, Lint and Test
+    name: Build and Test
     runs-on: ${{ inputs.runner || 'ubuntu-22.04' }}
+    env:
+      CACHE_PATH: cache
+      CI_REPO: EVerest/everest-core
+      CI_REPO_DIR: ci-repo
+      SOURCE_DIR: source
     steps:
       - name: Format branch name for cache key
         run: |
           BRANCH_NAME_FOR_CACHE="${GITHUB_REF_NAME//-/_}"
           echo "branch_name_for_cache=${BRANCH_NAME_FOR_CACHE}" >> "$GITHUB_ENV"
-      - name: Setup cache
-        uses: actions/cache@v3
+      - name: Restore cache
+        uses: actions/cache/restore@v3
+        id: restore-cache
         with:
-          path: cache
+          path: ${{ env.CACHE_PATH }}
           key: compile-${{ env.branch_name_for_cache }}-${{ github.sha }}
           restore-keys: |
             compile-${{ env.branch_name_for_cache }}-
             compile-
-      - name: Checkout everest-core
-        uses: actions/checkout@v3
+      - name: Checkout `${{ env.CI_REPO }}` to `${{ env.CI_REPO_DIR }}/` (for reusable workflow mode, see https://stackoverflow.com/a/74123003/1168315)
+        uses: actions/checkout@v4
         with:
-          path: source
-      - name: Run clang-format
-        uses: everest/everest-ci/github-actions/run-clang-format@v1.0.0
+          path: ${{ env.CI_REPO_DIR }}
+          repository: ${{ env.CI_REPO }}
+          ref: CI/reusable-workflow  # TODO get reference in general, see https://stackoverflow.com/questions/74784735 resp. https://github.com/actions/toolkit/issues/1264
+      - name: Checkout repo ${{ inputs.repository }} to `${{ env.SOURCE_DIR }}/`
+        uses: actions/checkout@v4
         with:
-          source-dir: source
-          extensions: hpp,cpp
-          exclude: cache
-      - name: Setup run scripts
-        run: |
-          mkdir scripts
-          rsync -a source/.ci/build-kit/ scripts
+          path: ${{ env.SOURCE_DIR }}
+          repository: ${{ inputs.repository }}
```

Comment:
This doesn't seem like the right way to set up a reusable workflow. When someone makes a fork, the CI instructions on their forked branches will point to whatever original repository is passed in, for example, [EVerest/libocpp](https://github.com/EVerest/libocpp/pull/356/files#diff-868425f4c6dd9927d83974d334ea46b9ec96854cdac9bb15b15f43bc4ad17820R9). What *should* happen is that the current branch that initialized the workflow should run, fetching the branch from the forked repository.

In the current state, this [fails at the checkout stage](https://github.com/US-JOET/libfsm/actions/runs/7660200585/job/20876978782#step:5:46) because we can't check out the original repository from our fork.

Reply:
I don't see the problem in the way the `reusable workflow` is implemented, but the way it is called.

In the[ example of `libocpp` ](https://github.com/EVerest/libocpp/pull/356/files#diff-868425f4c6dd9927d83974d334ea46b9ec96854cdac9bb15b15f43bc4ad17820R13)the parameter `repository: EVerest/libocpp` is used. You adapted it similiar in[ your example ](https://github.com/US-JOET/libfsm/actions/runs/7660200585/workflow#L12) with `repository: EVerest/libfsm`

To enable workflows running on forks I suggest the use of the[ env variables 
](https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables).

-> In both examples I would replace the line with `repository: ${{ github.repository }}`.

Reply:
Shouldn't the reusable workflow use that by default?

Reply:
yeah, maybe this parameter is obsolete, not sure about this.. I will spend some time on this in the next weeks to figure those things out